emacs/.emacs.d/config.org
#+TITLE: My Emacs Configuration
#+AUTHOR: Ali Bozorgzadeh
#+EMAIL: alixbozorgzadeh@gmail.com
#+STARTUP: indent overview :results none
#+OPTIONS: toc:2 email:t

# Useful links
# - https://panadestein.github.io/emacsd/#orgc03b4c8

* Important programs to load first

#+begin_src elisp
  (add-to-list 'load-path "~/.emacs.d/scripts/")

  (require 'elpaca-setup)  ;; The Elpaca Package Manager
  (require 'buffer-move)   ;; Buffer-move for better window management
#+end_src

* Disable Backup and Auto-save

#+begin_src elisp
  (setq make-backup-files nil)
  (setq auto-save-default nil)
#+end_src

* Indentation

#+begin_src elisp
  (setq-default fill-column 90)
  (setq-default indent-tabs-mode nil)
  (setq-default tab-width 4)
  (setq indent-line-function 'insert-tab)

  ;; Disable electric indent mode
  (electric-indent-mode -1)
  (add-hook 'after-change-major-mode-hook (lambda() (electric-indent-mode -1)))

  ;; Disable emacs' builtin indentation in text and fundamental mode (locally)
  (add-hook 'text-mode-hook #'(lambda () (electric-indent-local-mode -1)))
  (add-hook 'org-mode-hook #'(lambda () (electric-indent-local-mode +1)))
  (add-hook 'fundamental-mode-hook #'(lambda () (electric-indent-local-mode -1)))
#+end_src

* History and Buffer state

#+begin_src elisp
  (savehist-mode +1)
  ;; (setq savehist-additional-variables '(kill-ring search-ring regexp-search-ring))
  (setq savehist-file "~/.emacs.d/.savehist")

  (require 'saveplace)
  (setq save-place-file "~/.emacs.d/.saveplace")
  (save-place-mode +1)

  (recentf-mode 1)
  (setq recentf-max-menu-items 25)
  (setq recentf-max-saved-items 25)
  (setq-default recent-save-file "~/emacs.d/.recentf")

  ;; call `recentf-save-list` periodically, e.g. every 5 minutes
  (run-at-time nil (* 5 60) 'recentf-save-list)

  (global-auto-revert-mode 1)

#+end_src

* Fonts
#+begin_src elisp :results none
  (setq rc/font-family-mono "Iosevka NF")
  (setq rc/font-weight 'Regular)
  (setq rc/font-size-mono 200)

  (set-face-attribute 'default nil
    :font rc/font-family-mono
    :height rc/font-size-mono
    :weight rc/font-weight)
  ; (set-face-attribute 'variable-pitch nil
  ;   :font "Helvetica Neue"
  ;   :height 200
  ;   :weight 'medium)
  (set-face-attribute 'fixed-pitch nil
    :font rc/font-family-mono
    :height rc/font-size-mono
    :weight rc/font-weight)
  ;; Makes commented text and keywords italics.
  ;; This is working in emacsclient but not emacs.
  ;; Your font must have an italic face available.
  ;;(set-face-attribute 'font-lock-comment-face nil :slant 'italic)
  ;;(set-face-attribute 'font-lock-keyword-face nil :slant 'italic)

  ;; This sets the default font on all graphical frames created after restarting Emacs.
  ;; Does the same thing as 'set-face-attribute default' above, but emacsclient fonts
  ;; are not right unless I also add this method of setting the default font.
  ;;
  ;; NOTE: The use of concat inside the quoted list is incorrect because concat needs to be
  ;; evaluated. However, it's inside a quoted list '(...), which prevents its evaluation.
  ;; NOTE: You need to use a backquote ` (also known as quasi-quotation) with a comma to
  ;; allow the expressions to be evaluated. Here's the corrected version:
  (add-to-list 'default-frame-alist `(font . ,(concat
   rc/font-family-mono "-" (number-to-string (/ rc/font-size-mono 10.0)))))

  ;; Uncomment the following line if line spacing needs adjusting.
  (setq-default line-spacing 0.0) ;; 0.12

  ;; (add-text-properties (point-min) (point-max)
  ;;                      '(line-spacing 0.01 line-height 1.0))

  ;; (defun set-bigger-spacing ()
  ;;   (setq-local default-text-properties '(line-spacing 0.01 line-height 1.0)))
  ;; (add-hook 'text-mode-hook 'set-bigger-spacing)
  ;; (add-hook 'prog-mode-hook 'set-bigger-spacing)
#+end_src

* Theme
#+begin_src elisp
  ;; (use-package ef-themes
  ;;   :ensure t
  ;;   :demand t
  ;;   :init
  ;;   (load-theme 'modus-vivendi t))

  ;; (use-package catppuccin-theme :ensure t :demand t)

  ;; (use-package gruvbox-theme
  ;; :config
  ;; (load-theme 'gruvbox t))

  (use-package gruber-darker-theme
    :config
    (load-theme 'gruber-darker t))

  ;; (use-package zenburn-theme
  ;;   :config
  ;;   (load-theme 'zenburn t))

  ;; Fontify the whole line for headings (with a background color).
  ;; (setq org-fontify-whole-heading-line t)

  ;; (setq leuven-scale-outline-headlines nil)
  ;; (setq leuven-dark-scale-outline-headlines nil)

  ;; (load-theme 'adwaita t)

  ;; (use-package vscode-dark-plus-theme
  ;;   :config
  ;;   (load-theme 'vscode-dark-plus t)
  ;;   ;; (add-to-list 'default-frame-alist '(foreground-color . "#...."))
  ;;   (add-to-list 'default-frame-alist '(background-color . "#181818")))
#+end_src

* UI
#+begin_src elisp
  ;; start the initial frame maximized
  (add-to-list 'initial-frame-alist '(fullscreen . maximized))
  ;; start every frame maximized
  (add-to-list 'default-frame-alist '(fullscreen . maximized))

  ;; (menu-bar-mode 0)
  (setq inhibit-startup-message t)
  (tool-bar-mode -1)
  (tooltip-mode -1)
  ;; (scroll-bar-mode -1)
  (blink-cursor-mode -1)

  ;; Don't pop-up UI dialog when prompting
  (setq use-dialog-box nil)

  ;; Disable alarms
  (setq ring-bell-function 'ignore)
  (setq visible-bell 1)

  ;; Show column number
  (column-number-mode +1)

  ;; Relative line numbers
  (setq-default display-line-numbers-type 'relative)
  (add-hook 'prog-mode-hook 'display-line-numbers-mode)

  ;; ;; Display truncated lines (fold long lines
  ;; (toggle-truncate-lines nil)

  ;; Scroll only current line when truncating lines
  ;; src: https://emacs.stackexchange.com/a/40909/30552
  ;; (setq truncate-lines t) ; local mode
  (setq-default truncate-lines t) ; local mode
  ;; (setq-local auto-hscroll-mode 'current-line) ; major mode must be activated before this
  (setq-default auto-hscroll-mode 'current-line)
#+end_src

* Pairing parens and quotes
* COMMENT electric pair mode
#+begin_src elisp :results none
  ;; No delay when deleting pairs
  (setq-default delete-pair-blink-delay 0)
  (setq electric-pair-inhibit-predicate 'electric-pair-conservative-inhibit)
  (setq electric-pair-preserve-balance nil)
  (setq electric-pair-pairs
        '((8216 . 8217)
          (8220 . 8221)
          (171 . 187)))
  (setq electric-pair-skip-self 'electric-pair-default-skip-self)
  (setq electric-pair-skip-whitespace nil)
  (setq electric-pair-skip-whitespace-chars '(9 10 32))
  (setq electric-quote-context-sensitive t)
  (setq electric-quote-paragraph t)
  (setq electric-quote-string nil)
  (setq electric-quote-replace-double t)
  (electric-pair-mode 1)
#+end_src
** COMMENT Smart parens mode
#+begin_src elisp :results none
  (use-package smartparens
    :ensure smartparens  ;; install the package
    :diminish
    :hook (prog-mode text-mode markdown-mode) ;; add `smartparens-mode` to these hooks
    :config

    ;; Stolen from: https://github.com/iqbalansari/dotEmacs/blob/config/parenthesis.org

    ;; Custom colors for parenthesis
    (set-face-attribute 'show-paren-mismatch nil :foreground "white" :background "red")

    ;; Show matching paranthesis immediately
    (show-paren-mode t)

    (setq show-paren-when-point-inside-paren t)

    ;; Highlight only the paranthesis
    (setq show-paren-style 'parenthesis)  ; does not work but the following does effect
    (setq sp-highlight-pair-overlay nil)

    ;; Keybindings for navigating around sexps using smartparens
    (define-key sp-keymap (kbd "C-M-f") 'sp-forward-sexp)
    (define-key sp-keymap (kbd "C-M-b") 'sp-backward-sexp)
    (define-key sp-keymap (kbd "C-M-n") 'sp-next-sexp)
    (define-key sp-keymap (kbd "C-M-p") 'sp-previous-sexp)

    ;; load default config
    (require 'smartparens-config)
    (show-smartparens-global-mode 1)
    (smartparens-global-mode 1))
#+end_src

* Undo / Redo (UndoFu + Undo-tree)
#+begin_src elisp
  ;; (use-package undo-fu
  ;;   :diminish
  ;;   :config
  ;;   (global-set-key (kbd "M-[") 'undo-fu-only-undo)
  ;;   (global-set-key (kbd "M-]") 'undo-fu-only-redo))

  (use-package undo-tree
    :diminish
    :bind (("C-x C-u" . undo-tree-visualize)
           ("C-x u" . undo-tree-undo)
           ("M-[" . undo)
           ("M-]" . undo-tree-redo))
    :config
    ;; Prevent undo tree files from polluting your git repo
    (setq undo-tree-history-directory-alist '(("." . "~/.emacs.d/undo")))
    (global-undo-tree-mode))
#+end_src

* Org Mode
** Org settings

#+begin_src elisp :results none
  (use-package org
    :ensure nil
    :bind (("C-c a" . org-agenda))
    :config
    ;; (add-hook
    ;;  'org-mode-hook
    ;;  (lambda ()
    ;;    (setq-local
    ;;     electric-pair-inhibit-predicate
    ;;     `(lambda (c) (if (char-equal c ?<) t (,electric-pair-inhibit-predicate c))))))

    ;; Record time and note when a task is completed
    (setq org-log-done 'note)

    ;; Record time and note when the scheduled date of a task is modified
    (setq org-log-reschedule 'note)

    ;; Record time and note when the deadline of a task is modified
    (setq org-log-redeadline 'note)

    ;; Record time and note when clocking out of a task
    (setq org-log-clock-out 'note)

    ;; Record time and note when a task is refiled
    (setq org-log-refile 'note)

    ;; Record note when clocking out
    (setq org-log-note-clock-out t)

    ;; Log everything into the LOGBOOK drawer
    (setq org-log-into-drawer t)

    ;; Log inserting a heading
    ;; (setq org-trest-insert-todo-heading-as-state-change t)

    ;; Default TODO keywords
    (setq org-todo-keywords '((sequence "TODO(t)" "LATER(l)" "|" "DONE(d)" "CANCELED(c)")))

    ;; Agenda
    (setq org-agenda-files '("~/todo.txt"))

    ;;--- Theme settings
    ;; source: https://zzamboni.org/post/beautifying-org-mode-in-emacs/

    ;; (add-hook 'org-mode-hook 'fixed-pitch-mode)
    ;; (add-hook 'org-mode-hook 'visual-line-mode)

    (setq org-hide-emphasis-markers nil)

    ;; Use "bullets" instead of "-" in lists
    (font-lock-add-keywords 'org-mode
                            '(("^ *\\([-]\\) "
                               (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "-")))))) ; •
    )
#+end_src

<s'TAB'= expands to src block
#+begin_src elisp
  (require 'org-tempo)
#+end_src
** Org packages
#+begin_src elisp
  (use-package toc-org
    :diminish
    :commands toc-org-enable
    :init (add-hook 'org-mode-hook 'toc-org-enable))

  (use-package org-indent
    :ensure nil
    :diminish
    :config
    (add-hook 'org-mode-hook 'org-indent-mode)
    (use-package org-bullets)
    (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1))))
#+end_src
** Org capture
#+begin_src elisp
  (global-set-key (kbd "C-c c") 'org-capture)

  (setq org-directory "~/org")
  (setq org-default-notes-file (concat org-directory "/notes.org"))

  (setq rc/org-gtd-file (concat org-directory "/gtd.org"))
  (setq rc/org-journal-file (concat org-directory "/jrnl.org"))

  (setq org-capture-templates
        `(("t" "Todo"
           entry (file+headline ,rc/org-gtd-file "Tasks")
           "* TODO %? %^G\n:LOGBOOK:\n%U\n%a\n:END:\n%i" :prepend t)
          ("j" "Journal"
           entry (file+datetree ,rc/org-journal-file)
           "* %<%H:%M> %? %^G\n:LOGBOOK:\n%U\n%a\n:END:\n%i"
           :prepend t :empty-lines 1)
          ("c" "Collection"
           entry (file+headline ,rc/org-gtd-file "Collect")
           "* %? %^G \n  %^t")
          ))
#+end_src
** COMMENT Latex

Custom report class that does not equate 1st level heading to =\part{}=

#+begin_src elisp
  ;; (add-to-list 'org-latex-classes
  ;;   '("myreport"
  ;;     "\\documentclass[11pt]{report}"
  ;;     ;; ("\\part{%s}" . "\\part*{%s}")
  ;;     ("\\chapter{%s}" . "\\chapter*{%s}")
  ;;     ("\\section{%s}" . "\\section*{%s}")
  ;;     ("\\subsection{%s}" . "\\subsection*{%s}")
  ;;     ("\\subsubsection{%s}" . "\\subsubsection*{%s}")))
#+end_src

Using =latexmk=

#+begin_src elisp
  (setq
   org-latex-pdf-process
   (list
    "latexmk -pdf -dvi- -f %f"))
#+end_src

#+begin_src elisp
  (require 'ox-latex)
  (setq org-latex-pdf-process '("latexmk -pdf %f"))  ;; Customize your LaTeX build process as needed

  ;; Set citation export processors for LaTeX
  (setq org-cite-export-processors '((latex biblatex) (t csl)))

  ;; Optional: Set a default citation style for biblatex
  (setq org-cite-biblatex-options
        '(("style" . "numeric")))  ;; Use a numbered citation style
#+end_src

* Diminish
#+begin_src elisp
  (use-package diminish)
#+end_src
* Yasnippet

#+begin_src elisp
  (use-package yasnippet
    :ensure t
    :diminish yas-minor-mode
    :config
    (setq yas/triggers-in-field nil)
    (setq yas-snippet-dirs '("~/.emacs.snippets/"))
    (yas-global-mode 1))
#+end_src

* CompAny
#+begin_src elisp
  (use-package company
    :ensure t
    :diminish
    :delight
    :bind (("C-c ." . company-complete)
           ("C-c C-." . company-complete)
           ("C-c s s" . company-yasnippet)
           :map company-active-map
           ("C-n" . company-select-next)
           ("C-p" . company-select-previous)
           ("C-d" . company-show-doc-buffer)
           ("M-." . company-show-location))
    :init
    (add-hook 'c-mode-common-hook 'company-mode)
    (add-hook 'sgml-mode-hook 'company-mode)
    (add-hook 'emacs-lisp-mode-hook 'company-mode)
    (add-hook 'text-mode-hook 'company-mode)
    (add-hook 'lisp-mode-hook 'company-mode)
    :config
    (eval-after-load 'c-mode
      '(define-key c-mode-map (kbd "[tab]") 'company-complete))

    (setq company-tooltip-limit 20)
    (setq company-show-numbers t)
    (setq company-dabbrev-downcase nil)
    (setq company-idle-delay 0.3)
    (setq company-echo-delay 0.3)
    ;; (setq company-ispell-dictionary (f-join tychoish-config-path "aspell-pws"))

    (setq company-backends '(company-capf
                             company-keywords
                             company-semantic
                             company-files
                             company-etags
                             company-dabbrev ; for completion in the buffer (similar to M-/ hippie expand)
                             company-clang
                             company-cmake
                             company-yasnippet))

    (add-to-list 'company-backends '(company-keywords :with company-dabbrev))
    (add-to-list 'company-backends '(company-keywords :with company-dabbrev-code))

    (global-company-mode))

  (use-package company-quickhelp
   :after company
   :config
   (setq company-quickhelp-idle-delay 0.3)
   (company-quickhelp-mode 1))
#+end_src
* Compile Mode
#+begin_src elisp
  ;; Stole from https://github.com/curtainp/.emacs.d/tree/lisp/init-basic.el
  (use-package compile
    :ensure nil
    :hook (compilation-filter . ansi-color-compilation-filter) ; Enable ANSI colors in compilation buffer
    :hook (shell-mode . compilation-shell-minor-mode)
    :bind (("C-;" . compile)
           ("C-c C-;" . recompile))
    :custom
    (compilation-scroll-output t) ; Keep scrolling the compilation buffer, `first-error' can be interesting
    (compilation-always-kill t) ; Always kill current compilation process before starting a new one
    ;; (compilation-skip-visited t) ; Skip visited messages on compilation motion commands
    (compilation-window-height 12) ; Keep it readable  :init
    :config
    ;; Integration of `compile' with `savehist'
    (with-eval-after-load 'savehist
      (add-to-list 'savehist-additional-variables 'compile-history)))
#+end_src
* Expand Region
#+begin_src elisp
  (use-package expand-region
    :bind ("C-=" . er/expand-region))
#+end_src
* IDO (Interactively-Do) and Imenu
** Ido
#+begin_src elisp
  ;; https://www.masteringemacs.org/article/introduction-to-ido-mode
  (setq ido-enable-flex-matching t)
  (setq ido-everywhere t)
  (setq ido-use-filename-at-point 'guess)
  (setq ido-use-url-at-point -1)
  (setq ido-create-new-buffer 'always)
  (setq-default confirm-nonexistent-file-or-buffer nil)
  (setq ido-file-extensions-order '(".c" ".cpp" ".C" ".H" ".py" ".txt" ".md" ".lua" ".yml" ".toml" ".cfg"))
  (setq ido-ignore-directories '("build" ".cache" ".git" "__pycache__" "venv" "node_modules"))
  (setq completion-ignored-extensions '(".bak" ".tags" ".o" ".so" ".aux" ".log" ".a" ".out"))
  (setq ido-ignore-extensions t)
  (ido-mode 1)
#+end_src
** Ido-at-point
#+begin_src elisp
  (use-package ido-at-point
    :init
    (ido-at-point-mode))
#+end_src
** Idomenu
#+begin_src elisp
  (use-package idomenu
    :ensure t
    :bind ("C-c i" . idomenu))
#+end_src
** Imenu merging default index function
#+begin_src elisp
  ;; src: https://stackoverflow.com/a/21656063/13041067
  (defun rc/merge-imenu ()
    (interactive)
    (let ((mode-imenu (imenu-default-create-index-function))
          (custom-imenu (imenu--generic-function imenu-generic-expression)))
      (append mode-imenu custom-imenu)))
#+end_src
* SMEX
#+begin_src elisp
  (use-package smex
   :bind (("M-x" . smex))
   :init
    (setq smex-save-file "~/.emacs.d/.smex_save")
    (smex-initialize))
#+end_src
* Project

# #+begin_src elisp
#   (require 'project)
# #+end_src

Projectile

#+begin_src elisp
  (use-package projectile
    :diminish projectile-mode
    :init
    ;; (setq projectile-known-projects-file "~/.emacs.d/local-files/projectile-bookmarks.eld")
    ;; (setq projectile-cache-file "~/.emacs.d/local-files/projectile.cache")
    :config
    (projectile-global-mode)
    (setq projectile-project-search-path '(("~/code/projects/" . 1) ("~/code/opt/" . 1)))
    (setq projectile-completion-system 'default)
    (define-key projectile-mode-map (kbd "C-x C-p") 'projectile-command-map)
    :bind (("C-x p" . projectile-command-map)))
#+end_src
* Whitespace

#+begin_src elisp
  (use-package whitespace
    :ensure nil
    :diminish
    :config
    (setq whitespace-display-mappings
          ;; all numbers are Unicode codepoint in decimal. try (insert-char 182 ) to see it
          '(
            (space-mark 32 [183] [46]) ; 32 SPACE, 183 MIDDLE DOT 「·」, 46 FULL STOP 「.」
            (newline-mark 10 [182 10]) ; 10 LINE FEED
            (tab-mark 9 [187 9] [9655 9] [92 9]) ; 9 TAB, 9655 WHITE RIGHT-POINTING TRIANGLE 「▷」
            ))
    (setq whitespace-style '(face tabs trailing tab-mark trailing-mark))
    (setq show-trailing-whitespace t)

    ;; (defun whitespace-trailing-regexp (limit)
    ;;   "Match all trailing spaces. This overloads the definition in whitespace.el."
    ;;   (let ((status t))
    ;;     (while (unless (re-search-forward whitespace-trailing-regexp limit t)
    ;;              (setq status nil)))          ;; end of buffer
    ;;     status))

    ;; Don't enable whitespace for.
    (setq-default whitespace-global-modes
                  '(not shell-mode
                        help-mode
                        magit-mode
                        magit-diff-mode
                        ibuffer-mode
                        dired-mode
                        occur-mode))

    ;; Set whitespace-mode face to be always visible
    (add-hook 'prog-mode-hook 'whitespace-mode))
#+end_src

* Dired
#+begin_src elisp :results none
  (use-package dired
    :ensure nil
    :commands (dired dired-jump)
    :config
    ;; -agho
    ;; (setq dired-listing-switches "-agho --group-directories-first -F ")
    (setq dired-listing-switches "-alh")

    ;; Kill the current buffer when selecting a new directory.
    (setq dired-kill-when-opening-new-dired-buffer t)

    ;; Stole from tsoding
    (setq-default dired-dwim-target t)
    (setq dired-mouse-drag-files t))
#+end_src


** Tips
- When browsing in dired instead of hitting enter to see a directory use =i= then it adds
  that directory to the current buffer. ([[https://stackoverflow.com/posts/1839352/revisions][src]])

* Magit
Due to some complication in the installation of magit and transient the following is
needed

#+begin_src elisp :results none
  (setq package-install-upgrade-built-in t)

  (use-package transient
    :bind
    (:map transient-map
          ("<escape>" . transient-quit-all)))
#+end_src

Open the actual file when visiting symlink that points to a version-controlled file
#+begin_src elisp :results none
  (use-package vc
    :ensure nil
    :config
    (setq vc-follow-symlinks t))
#+end_src

MAGIT:
#+begin_src elisp :results none
  (use-package magit
    :ensure t
    :after transient
    :config
    (setq magit-save-repository-buffers nil) ; Disable Magit asking to save files
    (setq magit-uniquify-buffer-names nil)   ; Make magit buffers be wrapped w/ *
    (setq magit-auto-revert-mode nil)
    (global-set-key (kbd "C-c m s") 'magit-status)
    (global-set-key (kbd "C-c m l") 'magit-log)

    ;; Do not create a new window for magit buffers.
    (setq magit-display-buffer-function 'magit-display-buffer-same-window-except-diff-v1)
    (setq magit-bury-buffer-function 'magit-restore-window-configuration)

    ;; source: https://manueluberti.eu/posts/2018-02-17-magit-bury-buffer/
    (defun rc/magit-kill-buffers ()
      "Restore window configuration and kill all Magit buffers."
      (interactive)
      (let ((buffers (magit-mode-get-buffers)))
        (magit-restore-window-configuration)
        (mapc #'kill-buffer buffers)))
    (bind-key "q" #'rc/magit-kill-buffers magit-status-mode-map)

    ;; Custom functions for OPENING the REMOTE REPOSITORY

    (defun rc/parse-git-remote-url (url)
      "If necessary, convert an SSH to HTTPS git remote location."
      (if (string-match "^http" url)
          url
        (replace-regexp-in-string "\\(.*\\)@\\(.*\\):\\(.*\\)\\(\\.git?\\)"
                                  "https://\\2/\\3"
                                  url)))

    (defun rc/magit-open-remote-repo ()
      "Opens a remote repo URL. Prompts the user to choose a remote."
      (interactive)
      (let* ((remote-name (magit-read-remote "Choose remote repository"))
             (url (magit-get "remote" remote-name "url")))
        (browse-url (rc/parse-git-remote-url url))
        (message "Opening repo %s" url)))

    (defun rc/project-get-root-relative-path (&optional path)
      "Returns PATH relative to a project root."
      (let* ((path (expand-file-name (or path default-directory)))
             (root-path (expand-file-name
                         (cdr (project--find-in-directory path))))
             (root-length (length root-path))
             (rel-path (substring path root-length)))
        rel-path))

    (defun rc/magit-open-remote-dwim (&optional dir)
      "Opens a remote repo URL in the exact DIR location. Prompts the
  user to choose a remote."
      (interactive)
      (let* ((file (buffer-file-name))
             (dir (expand-file-name (or dir default-directory)))
             (remote-name (magit-read-remote "Choose remote repository" nil t))
             (url (rc/parse-git-remote-url
                   (magit-get "remote" remote-name "url")))
             ;; Identify remote hosting type (e.g., GitHub, GitLab, etc.)
             (remote-hosting-type
              (cond
               ((string-match-p (regexp-quote "github") url) "GitHub")
               ((string-match-p (regexp-quote "gitlab") url) "GitLab")
               (t (completing-read "Choose remote hosting type:"
                                   '("GitHub" "GitLab"))))))
        ;; Adapt to GitLab's URL standard
        (when (string= remote-hosting-type "GitLab")
          (setq url (concat url "/-")))
        ;; Adapt to differences in file and directory URL paths
        (if file
            (setq url (concat url "/blob/" (magit-get-current-branch)
                              "/" (rc/project-get-root-relative-path file)))
          (setq url (concat url "/tree/" (magit-getc-urrent-branch)
                            "/" (rc/project-get-root-relative-path dir))))
        ;; Perform actions
        (browse-url url)
        (message "Opening %s" url)))

    :commands
    (rc/magit-open-remote-repo
     rc/magit-open-remote-dwim))
#+end_src

Provide commit message guidelines/feedback when committing
#+begin_src elisp :results none
  ;; taken from https://protesilaos.com/dotemacs
  (use-package git-commit
    :after magit
    :custom
    (git-commit-fill-column 72)
    (git-commit-summary-max-length 50)
    (git-commit-known-pseudo-headers
     '("Signed-off-by"
       "Acked-by"
       "Modified-by"
       "Cc"
       "Suggested-by"
       "Reported-by"
       "Tested-by"
       "Reviewed-by"))
    (git-commit-style-convention-checks

nvim/.config/nvim/lua/config/options.lua
vim.cmd.colorscheme 'vscode'

-- [ Line Numbers ] --------------------------


vim.opt.number         = true
vim.opt.relativenumber = true


-- [ Indentation ] --------------------------

vim.opt.autoindent = true -- maintain indent of current line
vim.opt.expandtab  = true -- always use tabs instead of spaces
vim.opt.smarttab   = true -- <tab>/<BS> indent/deindent in loading whitespace
vim.opt.shiftwidth = 2
vim.opt.tabstop    = 2 -- spaces per tab


-- [ Search ] --------------------------

vim.opt.hlsearch   = true
vim.opt.ignorecase = true
vim.opt.incsearch  = true
vim.opt.smartcase  = true


-- [ Scroll ] --------------------------

vim.opt.scrolloff     = 8 -- start scroll 3 lines before edge of the viewport
vim.opt.sidescrolloff = 3


-- [ Splits ] --------------------------

vim.opt.splitbelow = true -- open horizontal splits below current window
vim.opt.splitright = true -- open vertical splits to the right of the current window


-- [ Backup ] --------------------------

vim.opt.backup      = false
vim.opt.swapfile    = false -- don't create swap file
vim.opt.writebackup = false


-- [ Path ] --------------------------

vim.opt.path = "**"
-- vim.opt.path = vim.opt.path + vim.fs.normalize("$HOME/.config/nvim")


-- [ Wild Menu ] --------------------------

vim.opt.wildmenu       = true -- show options as list when switching to buffers etc
vim.opt.wildignorecase = true
vim.opt.wildchar       = 9 -- ('<Tab>') character to trigger wildcard expansion
vim.opt.wildcharm      = 26 -- ('<C-z>') substitue for 'wildchar' in macros

-- shell-like autocomplete to unambigious portion
vim.opt.wildmode = {
    'longest', 'list', 'full'
}

-- patterns to ingore durng file navigation
vim.opt.wildignore:append({
    '*.git/*', '*.tags', 'tags', '*build/*', '*.o', '*.class', '*.egg-info/**',
    '__pycache__', 'venv', '*venv/**', '*dist/**', '**/.vim/undo/**', '**/.vim/pack/**'
})


-- [ Undo-Files ] --------------------------

if root then
    vim.opt.undofile = false -- Don't create root-owned files
else
    vim.opt.undofile = true
    -- Keep undo file out of the way
    -- (Check if the undodir exist. If not create one)
    local undo_dir = "$XDG_CACHE_HOME/nvim/undo/"
    if not vim.fn.isdirectory(undo_dir) then
        print("No undo directory. Creating one in " .. undo_dir .. ".")
        vim.fn.system("mkdir -p " .. undo_dir)
    end
    vim.opt.undodir = vim.fn.expand(undo_dir)
    vim.opt.undodir:append(".") -- fallback
end


-- [ Special Characters ] --------------------------

vim.opt.fillchars = {
    diff = "∙", -- BULLET OPERATOR (U+2219, UTF-8: E2 88 99)
    eob  = " ", -- NO-BREAK SPACE (U+00A0, UTF-8: C2 A0) to suppress ~ at EndOfBuffer
    fold = " ", -- Space
    vert = "┃", -- BOX DRAWINGS HEAVY VERTICAL (U+2503, UTF-8: E2 94 83)
}

vim.opt.list = true
vim.opt.listchars = {
    nbsp     = '¬', -- NOT SIGN (U+00AC, UTF-8: C2 AC)
    extends  = '»', -- RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK (U+00BB, UTF-8: C2 BB)
    precedes = '«', -- LEFT-POINTING DOUBLE ANGLE QUOTATION MARK (U+00AB, UTF-8: C2 AB)
    tab      = '▸ ', -- BLACK RIGHT-POINTING SMALL TRIANGLE (U+25B8, UTF-8: E2 96 B8)
    trail    = '•', -- BULLET (U+2022, UTF-8: E2 80 A2)
}


-- [ Misc. ] --------------------------

vim.opt.showcmd = true
vim.opt.showbreak = '↪ ' -- ↳ DOWNWARDS ARROW WITH TIP RIGHTWARDS (U+21B3, UTF-8: E2 86 B3)
vim.opt.linebreak = true
vim.opt.cmdheight   = 1
vim.opt.laststatus  = 3
-- vim.opt.ruler       = true
-- vim.opt.rulerformat = [[%30(%{FugitiveStatusline()}%=%l,%c%V   %P%)]]
-- vim.opt.statusline = [[%f %y %m %r%=%30(%{FugitiveStatusline()}  %l,%c   %P%)]]
vim.opt.statusline = [[%f %y %m %r%=%30(%l,%c   %P%)]]
vim.opt.textwidth   = 81 -- automatically hard wrap at 120 columns
vim.opt.updatetime  = 250 -- CursorHold interval

vim.opt.pumheight = 10

vim.opt.wrap          = false
vim.opt.cursorline    = false
vim.opt.equalalways   = false
vim.opt.joinspaces    = false -- don't autoinsert two spaces after '.', '?', '!' for join command
vim.opt.lazyredraw    = false -- don't bother updating screen during macro playback
vim.opt.showmatch     = false
vim.opt.showmode      = false
vim.opt.termguicolors = true
vim.opt.title         = true

-- vim.opt.guicursor       = ""
vim.opt.encoding    = "utf-8"
vim.opt_local.fileformat  = "unix"
-- vim.opt.foldmethod      = 'indent'
vim.opt.inccommand  = "split" -- live preview of :s results
vim.opt.mouse       = "a" -- enable mouse
vim.opt.shortmess   = "filnxtToOc"
vim.opt.signcolumn  = "yes"
vim.opt.virtualedit = "block" -- allow cursor to move where there is no text in visual block mode
vim.opt.spellcapcheck = '' -- don't check for capital letters at start of sentence

if vim.loop.os_uname().sysname == "Darwin" then
vim.opt.shell       = "zsh"
else
  vim.opt.shell     = "bash"
end

vim.opt.clipboard:append("unnamedplus")
vim.opt.isfname:remove("=") -- Avoid recognizing = as part of file name
vim.opt.iskeyword:append("-") -- treat dash separated words as a word text object

vim.opt.completeopt = {
    "menu","menuone", "noselect"
}

vim.opt.fileencodings = {
    "utf-8"
}

vim.opt.suffixesadd = {
    ".js", ".es", ".jsx", ".json", ".css", ".less", "sass", ".styl", ".php",
    ".py", ".md"
}

-- Focused
local focused_colorcolumn = '+' .. table.concat({
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',
    '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23',
    '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34',
    '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45',
    '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56',
    '57', '58', '59', '60', '61', '62', '63', '64', '65', '66', '67',
    '68', '69', '70', '71', '72', '73', '74', '75', '76', '77', '78',
    '79', '80', '81', '82', '83', '84', '85', '86', '87', '88', '89',
    '90', '91', '92', '93', '94', '95', '96', '97', '98', '99', '100',
    '101', '102', '103', '104', '105', '106', '107', '108', '109', '110',
    '111', '112', '113', '114', '115', '116', '117', '118', '119', '120',
    '121', '122', '123', '124', '125', '126', '127', '128', '129', '130',
    '131', '132', '133', '134', '135', '136', '137', '138', '139', '140',
    '141', '142', '143', '144', '145', '146', '147', '148', '149', '150',
    '151', '152', '153', '154', '155', '156', '157', '158', '159', '160',
    '161', '162', '163', '164', '165', '166', '167', '168', '169', '170',
    '171', '172', '173', '174', '175', '176', '177', '178', '179', '180',
    '181', '182', '183', '184', '185', '186', '187', '188', '189', '190',
    '191', '192', '193', '194', '195', '196', '197', '198', '199', '200',
    '201', '202', '203', '204', '205', '206', '207', '208', '209', '210',
    '211', '212', '213', '214', '215', '216', '217', '218', '219', '220',
    '221', '222', '223', '224', '225', '226', '227', '228', '229', '230',
    '231', '232', '233', '234', '235', '236', '237', '238', '239', '240',
    '241', '242', '243', '244', '245', '246', '247', '248', '249', '250',
    '251', '252', '253', '254'
}, ',+')
vim.wo.colorcolumn = focused_colorcolumn
